File: lista_1/c_library/iterative_lib.c
#include "iterative_lib.h"
#include <stdlib.h>

// Implementacja funkcji iteracyjnej obliczającej silnię
long long factorial_iterative(int n) {
    if (n < 0)
        return -1;
    long long result = 1;
    for (int i = 2; i <= n; ++i) {
        result *= i;
    }
    return result;
}

// Implementacja funkcji iteracyjnej obliczającej NWD
unsigned int gcd_iterative(unsigned int a, unsigned int b) {
    while (b != 0) {
        unsigned int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Implementacja funkcji iteracyjnej rozwiązującej równanie diofantyczne
Solution diophantine_iterative(int a, int b, int c) {
    Solution sol = {0, 0, false};
    
    int gcd_ab = gcd_iterative(a, b);
    if (c % gcd_ab != 0) {
        // Brak rozwiązania, gdy c nie jest podzielne przez NWD(a, b)
        sol.err = true;
        return sol;
    }
    
    // Obliczenie współczynników x i y przy pomocy rozszerzonego algorytmu Euklidesa
    int x = 1, y = 0;
    int x1 = 0, y1 = 1;
    int temp;
    
    while (b != 0) {
        int q = a / b;
        int r = a % b;
        
        temp = x1;
        x1 = x - q * x1;
        x = temp;
        
        temp = y1;
        y1 = y - q * y1;
        y = temp;
        
        a = b;
        b = r;
    }
    
    // Obliczenie rozwiązania dla równania diofantycznego
    int k = c / gcd_iterative(a, b);
    sol.x = x * k;
    sol.y = y * k;
    
    return sol;
}

File: lista_1/python_library/iterative_lib.py
# iterative_lib.py
from struct_lib import Solution


# Funkcja obliczająca silnię iteracyjnie
def factorial_iterative(n):
    if n < 0:
        return -1
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result


# Funkcja obliczająca NWD iteracyjnie
def gcd_iterative(a, b):
    while b != 0:
        a, b = b, a % b
    return a


# Funkcja rozwiązująca równanie diofantyczne iteracyjnie
def diophantine_iterative(a, b, c):
    sol = Solution(0, 0, False)
    
    gcd_ab = gcd_iterative(a, b)
    if c % gcd_ab != 0:
        # Brak rozwiązania, gdy c nie jest podzielne przez NWD(a, b)
        sol.err = True
        return sol
    
    # Obliczenie współczynników x i y przy pomocy rozszerzonego algorytmu Euklidesa
    x, y = 1, 0
    x1, y1 = 0, 1
    
    while b != 0:
        q = a // b
        r = a % b
        
        x, x1 = x1, x - q * x1
        y, y1 = y1, y - q * y1
        
        a, b = b, r
    
    # Obliczenie rozwiązania dla równania diofantycznego
    k = c // gcd_iterative(a, b)
    sol.x = x * k
    sol.y = y * k
    
    return sol
